File: 068_WordSearch.cpp
// Topic: Word Search
// Description: Determine if a word exists in grid sequentially (adjacent cells, no reuse).
// Approach: Backtracking DFS with visited marking.

#include <bits/stdc++.h>
using namespace std;

class Solution
{
   public:
    bool exist(vector<vector<char>>& board, const string& word)
    {
        int m = board.size();
        if (m == 0)
            return false;
        int n = board[0].size();
        for (int i = 0; i < m; ++i)
            for (int j = 0; j < n; ++j)
                if (board[i][j] == word[0])
                {
                    if (dfs(board, i, j, word, 0))
                        return true;
                }
        return false;
    }

   private:
    bool dfs(vector<vector<char>>& b, int i, int j, const string& w, int idx)
    {
        if (idx == (int)w.size())
            return true;
        int m = b.size(), n = b[0].size();
        if (i < 0 || j < 0 || i >= m || j >= n || b[i][j] != w[idx])
            return false;
        char tmp = b[i][j];
        b[i][j] = '#';
        bool found = dfs(b, i - 1, j, w, idx + 1) || dfs(b, i + 1, j, w, idx + 1) ||
                     dfs(b, i, j - 1, w, idx + 1) || dfs(b, i, j + 1, w, idx + 1);
        b[i][j] = tmp;
        return found;
    }
};

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    vector<vector<char>> board = {{'A', 'B', 'C', 'E'}, {'S', 'F', 'C', 'S'}, {'A', 'D', 'E', 'E'}};
    cout << "Test 1 (ABCCED) -> " << (Solution().exist(board, string("ABCCED")) ? "true" : "false")
         << "\n";
    cout << "Test 2 (SEE) -> " << (Solution().exist(board, string("SEE")) ? "true" : "false")
         << "\n";
    cout << "Test 3 (ABCB) -> " << (Solution().exist(board, string("ABCB")) ? "true" : "false")
         << "\n";
}

/*
Time Complexity: O(m*n * 4^L) worst-case where L is word length
Space Complexity: O(L) recursion depth
*/

File: 068_WordSearch.py
# Topic: Word Search
# Description: Determine if a word exists in grid sequentially (adjacent cells, no reuse).
# Approach: Backtracking DFS with visited marking.

from typing import List


class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        m, n = len(board), len(board[0])

        def dfs(i, j, idx):
            if idx == len(word):
                return True
            if i < 0 or j < 0 or i >= m or j >= n or board[i][j] != word[idx]:
                return False
            tmp = board[i][j]
            board[i][j] = '#'
            found = (dfs(i - 1, j, idx + 1) or dfs(i + 1, j, idx + 1) or
                     dfs(i, j - 1, idx + 1) or dfs(i, j + 1, idx + 1))
            board[i][j] = tmp
            return found

        for i in range(m):
            for j in range(n):
                if board[i][j] == word[0] and dfs(i, j, 0):
                    return True
        return False


if __name__ == "__main__":
    board = [['A', 'B', 'C', 'E'], ['S', 'F', 'C', 'S'], ['A', 'D', 'E', 'E']]
    print(f"Test 1 (ABCCED) -> {Solution().exist(board, 'ABCCED')}")
    print(f"Test 2 (SEE) -> {Solution().exist(board, 'SEE')}")
    print(f"Test 3 (ABCB) -> {Solution().exist(board, 'ABCB')}")

"""
Time Complexity: O(m*n * 4^L) worst-case where L is word length
Space Complexity: O(L) recursion depth
"""
